lib/digital_signal/digital_signal.c

```diff
@@ -80,94 +179,480 @@ uint32_t digital_signal_get_edge(DigitalSignal* signal, uint32_t edge_num) {
-    LL_DMA_SetDataLength(DMA1, LL_DMA_CHANNEL_2, signal->edge_cnt - 2);
-    LL_DMA_EnableChannel(DMA1, LL_DMA_CHANNEL_2);
+static void digital_signal_stop_dma() {
+    LL_DMA_DisableChannel(DMA1, LL_DMA_CHANNEL_1);
+    LL_DMA_DisableChannel(DMA1, LL_DMA_CHANNEL_2);
+    LL_DMA_ClearFlag_TC1(DMA1);
+    LL_DMA_ClearFlag_TC2(DMA1);
+}
+
+static void digital_signal_stop_timer() {
+    LL_TIM_DisableCounter(TIM2);
+    LL_TIM_DisableUpdateEvent(TIM2);
+    LL_TIM_DisableDMAReq_UPDATE(TIM2);
+}
+
+static void digital_signal_setup_timer() {
+    digital_signal_stop_timer();
 
-    // Set up timer
     LL_TIM_SetCounterMode(TIM2, LL_TIM_COUNTERMODE_UP);
     LL_TIM_SetClockDivision(TIM2, LL_TIM_CLOCKDIVISION_DIV1);
     LL_TIM_SetPrescaler(TIM2, 0);
-    LL_TIM_SetAutoReload(TIM2, 10);
+    LL_TIM_SetAutoReload(TIM2, 0xFFFFFFFF);
     LL_TIM_SetCounter(TIM2, 0);
+}
+
+static void digital_signal_start_timer() {
+    LL_TIM_EnableCounter(TIM2);
     LL_TIM_EnableUpdateEvent(TIM2);
     LL_TIM_EnableDMAReq_UPDATE(TIM2);
+    LL_TIM_GenerateEvent_UPDATE(TIM2);
+}
 
-    // Start transactions
-    LL_TIM_GenerateEvent_UPDATE(TIM2); // Do we really need it?
-    LL_TIM_EnableCounter(TIM2);
+static bool digital_signal_setup_dma(DigitalSignal* signal) {
+    furi_assert(signal);
+    DigitalSignalInternals* internals = signal->internals;
 
-    while(!LL_DMA_IsActiveFlag_TC2(DMA1))
-        ;
+    if(!signal->internals->reload_reg_entries) {
+        return false;
+    }
+    digital_signal_stop_dma();
 
-    LL_DMA_ClearFlag_TC1(DMA1);
-    LL_DMA_ClearFlag_TC2(DMA1);
-    LL_TIM_DisableCounter(TIM2);
-    LL_TIM_SetCounter(TIM2, 0);
-    LL_DMA_DisableChannel(DMA1, LL_DMA_CHANNEL_1);
-    LL_DMA_DisableChannel(DMA1, LL_DMA_CHANNEL_2);
+    internals->dma_config_gpio.MemoryOrM2MDstAddress = (uint32_t)internals->gpio_buff;
+    internals->dma_config_gpio.PeriphOrM2MSrcAddress = (uint32_t) & (internals->gpio->port->BSRR);
+    internals->dma_config_timer.MemoryOrM2MDstAddress = (uint32_t)signal->reload_reg_buff;
+    internals->dma_config_timer.NbData = signal->internals->reload_reg_entries;
+
+    /* set up DMA channel 1 and 2 for GPIO and timer copy operations */
+    LL_DMA_Init(DMA1, LL_DMA_CHANNEL_1, &internals->dma_config_gpio);
+    LL_DMA_Init(DMA1, LL_DMA_CHANNEL_2, &internals->dma_config_timer);
+
+    /* enable both DMA channels */
+    LL_DMA_EnableChannel(DMA1, LL_DMA_CHANNEL_1);
+    LL_DMA_EnableChannel(DMA1, LL_DMA_CHANNEL_2);
+
+    return true;
+}
+
+void digital_signal_send(DigitalSignal* signal, const GpioPin* gpio) {
+    furi_assert(signal);
+
+    if(!signal->edge_cnt) {
+        return;
+    }
+
+    /* Configure gpio as output */
+    signal->internals->gpio = gpio;
+    furi_hal_gpio_init(
+        signal->internals->gpio, GpioModeOutputPushPull, GpioPullNo, GpioSpeedVeryHigh);
+
+    /* single signal, add a temporary, terminating edge at the end */
+    signal->edge_timings[signal->edge_cnt++] = 10;
```

Comment:
@g3gg0 I thing the problem is here. For Mf Classic emulation this extra edge leads to incorrect pulse_duration set. Without this line emulation works much better.
Why do you add this?

Reply:
IIRC this was due to the tests assumed there is a trailing edge.
see the discussion there:
https://github.com/flipperdevices/flipperzero-firmware/pull/2070#issuecomment-1344861543
so i added a small edge, causing the polarity to switch

Reply:
I wouldn't imagine so if the rest of the ISO15693 code still works (my picopass code doesn't go that low level). Happy to remove it for now and we can re-test

Reply:
i also found only one location where this function is used, which is the [furi_hal_nfc.c](https://github.com/flipperdevices/flipperzero-firmware/blob/0a3ff7f85ae653629957dfdd970a6fd3310ad14e/firmware/targets/f7/furi_hal/furi_hal_nfc.c) code.
so i **agree** in removing this trailing edge.

i hope the 10 ticks are not troublesome due to the timing get messed up (delaying the next signal)?